src/hv_anndata/interface.py

```diff
@@ -73,40 +44,333 @@ class AnnDataInterface(hv.core.Interface):
+                obs_selections[k[7:].replace("']", "")] = v
+            elif k.startswith("var."):
+                var_selections[k[4:]] = v
+            elif k.startswith("A.var['"):
+                var_selections[k[7:].replace("']", "")] = v
+        if obs_selections:
+            obs = Dataset(dataset.data.obs).select(**obs_selections).data.index
+        else:
+            obs = slice(None)
+        if var_selections:
+            var = Dataset(dataset.data.var).select(**var_selections).data.index
+        else:
+            var = slice(None)
+        return dataset.data[obs, var]
 
     @classmethod
     def values(
         cls,
-        data: hv.Dataset,
-        dim: hv.Dimension | str,
+        data: Dataset,
+        dim: Dimension | str,
         expanded: bool = True,  # noqa: FBT001, FBT002, ARG003
-        flat: bool = True,  # noqa: FBT001, FBT002, ARG003
+        flat: bool = True,  # noqa: FBT001, FBT002
         *,
         compute: bool = True,  # noqa: ARG003
-        keep_index: bool = False,  # noqa: ARG003
-    ) -> np.ndarray:
+        keep_index: bool = False,
+    ) -> np.ndarray | pd.api.extensions.ExtensionArray:
         """Retrieve values for a dimension."""
-        dim = data.get_dimension(dim)
-        proxy = cast("_AnnDataProxy", data.data)
-        return proxy[dim.name]
+        dim = cast("AdPath", data.get_dimension(dim))
+        adata = cast("AnnData", data.data)
+        values = dim(adata)
+        if not keep_index and isinstance(values, pd.Series):
+            values = values.array
+        elif flat and values.ndim > 1:
+            assert not isinstance(values, pd.api.extensions.ExtensionArray)  # noqa: S101
+            values = values.flatten()
+        return values
 
     @classmethod
-    def dimension_type(cls, data: hv.Dataset, dim: hv.Dimension | str) -> np.dtype:
+    def dtype(
+        cls, data: Dataset, dim: Dimension | str
+    ) -> np.dtype | pd.api.extensions.ExtensionDtype:
         """Get the data type for a dimension."""
-        dim = data.get_dimension(dim)
-        proxy = cast("_AnnDataProxy", data.data)
-        return proxy[dim.name].dtype
+        dim = cast("AdPath", data.get_dimension(dim))
+        adata = cast("AnnData", data.data)
+        return dim(adata).dtype
+
+    @classmethod
+    def dimension_type(cls, dataset: Dataset, dim):
+        return cls.dtype(dataset, dim).type
+
+    @classmethod
+    def iloc(cls, dataset: Dataset, index):
+        rows, cols = index
+        axes = cls.axes(dataset)
+        if cols != slice(None):
+            msg = (
+                f"When indexing using .iloc on {axes[0]} indexed data you may only select "
+                "rows along that dimension, i.e. you may not provide a column selection. "
+            )
+            raise IndexError(msg)
+        if axes[0] == "var":
+            return dataset.data[:, rows]
+        if axes[0] == "obs":
+            return dataset.data[rows]
```

Comment:
what’s the type of `index` in `iloc`? can it be more than just `(:, int(?)) | (int(?), :)`?

what happens when this code runs off the end here? there is no else branch or `assert`

Reply:
Definitely needs to handle other conditions. Generally in HoloViews it supports scalars, slice and sequences.